### PR TITLE
editor-selector-menu: split into submenus

### PR DESCRIPTION
**Pull request by helenphan0**
screen shot 2017-08-11 at 3 04 49 pm

- for tooltips and hotspots, there are no categories with submenus

### CODE REVIEW

**Comment by GuruM** [comment]
! Could we do specific icons for each category and then put a right pointing caret icon on the right?

**Comment by trisweb** [comment]
Awesome! Exactly what guru said.

Mockup:

screenshot 2017-08-11 12 20 08

**Comment by chelliah** [comment]
Nicee! Couple quick initial comments. For categories without any submenu items (Forms in the instance of this gif) we probably don't even need to render that category. Also, I think the submenus should be aligned with their corresponding category, so that you can always move the mouse over to them (I couldn't get to the media submenu in this case). We could also see how it would look if we just used the submenu item (Images) if there was only one submenu item in a category

File: src/helpers/constants.js [hunk 1]
```diff
@@ -8,6 +12,37 @@ export const GALLERY_TYPES = {
...
+    content: 'advancedElements',
+    icon: EllipsisButton,
+    willExpand: true
+  },
+]
+
+export const textEditors = ['RichText']
+export const formEditors = ['TextInput', 'TextAreaInput', 'SelectionField', 'Rating'];
+export const mediaEditors = ['Image', 'Hero', 'Video'];
+export const advancedEditors = ['Button', 'HTML'];
```
Comment (GuruM) [inline src/helpers/constants.js]: can we just put all these into a `items` attribute in the `categories` array above? Then we can remove the `willExpand` and just check if the length of `.items` is `> 1`

File: src/helpers/constants.js [hunk 2]
```diff
@@ -8,6 +12,37 @@ export const GALLERY_TYPES = {
...
+    icon: FormButton,
+    willExpand: true
+  },
+  {
+    name: 'Advanced',
+    content: 'advancedElements',
+    icon: EllipsisButton,
+    willExpand: true
+  },
+]
```
Comment (GuruM) [inline src/helpers/constants.js]: ;

File: src/icons/MediaButton.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import React from 'react';
+
+import IconButton from './IconButton';
+
+export default class AdvancedStyling extends React.PureComponent {
```
Comment (GuruM) [inline src/icons/MediaButton.js]: class name

File: src/icons/EllipsisButton.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import React from 'react';
+
+import IconButton from './IconButton';
+
+export default class AdvancedStyling extends React.PureComponent {
```
Comment (GuruM) [inline src/icons/EllipsisButton.js]: class name

File: src/icons/RightButton.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import React from 'react';
+
+import IconButton from './IconButton';
+
+export default class AdvancedStyling extends React.PureComponent {
```
Comment (GuruM) [inline src/icons/RightButton.js]: class name

File: src/components/EditorSelector.js [hunk 6]
```diff
@@ -105,13 +116,34 @@ export default class EditorSelector extends React.Component {
...
+        return (mediaEditors.includes(editor.type) && allowedEditorTypes.includes(editor.type))
+    })
+    editorUpdate.formElements = !allowedEditorTypes.isEmpty() && editors.filter((editor) => {
+        return (formEditors.includes(editor.type) && allowedEditorTypes.includes(editor.type))
+    })
+    editorUpdate.advancedElements = !allowedEditorTypes.isEmpty() && editors.filter((editor) => {
+        return (advancedEditors.includes(editor.type) && allowedEditorTypes.includes(editor.type))
+    })
+
+    this.setState(editorUpdate)
```
Comment (GuruM) [inline src/components/EditorSelector.js]: probably dont need any new code in this function at all if we make the changes in the `categories` constant

File: src/components/EditorSelector.js [hunk 7]
```diff
@@ -105,13 +116,34 @@ export default class EditorSelector extends React.Component {
     this.state = {
       showForm: false,
       primaryHoverMenu: '',
-      secondaryMenuHover: '',
+      menuState: '',
```
Comment (GuruM) [inline src/components/EditorSelector.js]: wasn't sure what this was until I went digging, maybe something like `openSubMenu`

File: src/components/EditorSelector.js [hunk 8]
```diff
@@ -143,99 +175,104 @@ export default class EditorSelector extends React.Component {
 
     const secondaryMenuStyle = {
       zIndex: 101,
-      width: 180,
+      width: 160,
       position: 'absolute',
-      top: 0,
-      left: 130
+      bottom: '0',
```
Comment (GuruM) [inline src/components/EditorSelector.js]: doesn't need to be a string

File: src/helpers/constants.js [hunk 9]
```diff
@@ -8,6 +12,33 @@ export const GALLERY_TYPES = {
   EMOJI: 'emoji'
 };
 
+export const categories = [
+  {
+    name: 'Text',
+    content: 'textElements'
```
Comment (GuruM) [inline src/helpers/constants.js]: why can't the content be the array itself

Comment (helenphan0) [inline src/helpers/constants.js]: Still working on this. The logic needs to work for Tooltips/Hotspots too, which have RichText, Image, and HTML editors.

Comment (GuruM) [inline src/helpers/constants.js]: 👍  the "content.length == 1" stuff should make it work out of the box I think

File: src/helpers/constants.js [hunk 10]
```diff
@@ -8,6 +12,41 @@ export const GALLERY_TYPES = {
...
+  TEXTAREAINPUT: 'TextAreaInput',
+  RADIO: 'SelectionField',
+  RATING: 'Rating',
+  BUTTON: 'Button'
+}
+
+export const categories = [
+  {
+    name: 'Text',
+    content: ['RichText']
```
Comment (GuruM) [inline src/helpers/constants.js]: since we have constants, we can use em instead of strings

File: src/components/EditorSelector.js [hunk 11]
```diff
@@ -143,99 +149,119 @@ export default class EditorSelector extends React.Component {
...
-              </a>
-            );
-          })}
+          { 
+            editorMenu.map((category) => {
+              if (category.items.length === 1) {
+                return this.renderSubMenuItems(category.items)
+              }
+              else {
+                return ( category.items.length ?
```
Comment (GuruM) [inline src/components/EditorSelector.js]: `category.items.length &&`

File: src/components/EditorSelector.js [hunk 12]
```diff
@@ -143,99 +149,119 @@ export default class EditorSelector extends React.Component {
...
-                    cursor="pointer"
-                  />
-                </div>
-
-              </a>
-            );
-          })}
+          { 
+            editorMenu.map((category) => {
+              if (category.items.length === 1) {
```
Comment (GuruM) [inline src/components/EditorSelector.js]: complex maps with conditionals are a natural place to break off components

**Comment by GuruM** [comment]
LGTM! 🇿🇼 the tidying up can also be done in a separate PR if we make a card for it, up to you!

File: src/components/EditorSelector.js (hunk 11 above)
Comment (helenphan0) [inline src/components/EditorSelector.js]: screen shot 2017-08-14 at 11 57 20 pm

I get 0 with &&

Comment (GuruM) [inline src/components/EditorSelector.js]: ah, you're right

**Pull request merged**